spring-session/src/main/java/org/springframework/session/jdbc/JdbcOperationsSessionRepository.java

```diff
@@ -0,0 +1,483 @@
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+
+/**
+ * A {@link org.springframework.session.SessionRepository} implementation that uses
+ * Spring's {@link JdbcOperations} to store sessions in a relational database. This
+ * implementation does not support publishing of
+ * {@link org.springframework.session.events.SessionDestroyedEvent} and its subclasses.
+ * <p>
+ * An example of how to create a new instance can be seen below:
+ *
+ * <pre class="code">
+ * JdbcTemplate jdbcTemplate = new JdbcTemplate();
+ *
+ * JdbcOperationsSessionRepository sessionRepository = new JdbcOperationsSessionRepository(jdbcTemplate);
+ * </pre>
+ *
+ * For additional information on how to create and configure a JdbcTemplate, refer to the
+ * <a href="http://docs.spring.io/spring/docs/current/spring-framework-reference/html/">
+ * Spring Framework Reference Documentation</a>.
+ * <p>
+ * By default, this implementation uses <code>SPR_SESSION</code> table to store sessions.
+ * SQL scripts to initialize the default schema are packaged as <code>schema-*.sql</code>,
+ * where <code>*</code> is the target database type. Note that the table prefix can be
+ * customized using the {@link #setTablePrefix(String)} method.
+ *
+ * @author Vedran Pavic
+ * @since 1.2.0
+ */
+public class JdbcOperationsSessionRepository
+		implements FindByIndexNameSessionRepository<JdbcOperationsSessionRepository.JdbcSession>, ApplicationEventPublisherAware {
+
+	private static final Log logger = LogFactory.getLog(JdbcOperationsSessionRepository.class);
+
+	private static final String DEFAULT_TABLE_PREFIX = "SPR_";
+
+	private static final String SPRING_SECURITY_CONTEXT = "SPRING_SECURITY_CONTEXT";
+
+	private static final PrincipalNameResolver PRINCIPAL_NAME_RESOLVER = new PrincipalNameResolver();
+
+	private enum Query {
+
+		CREATE_SESSION(
+				"INSERT INTO %PREFIX%SESSION(SESSION_ID, LAST_ACCESS_TIME, PRINCIPAL_NAME, SESSION_BYTES) " +
+						"VALUES (?, ?, ?, ?)"),
+
+		GET_SESSION(
+				"SELECT SESSION_BYTES FROM %PREFIX%SESSION WHERE SESSION_ID = ?"),
+
+		UPDATE_SESSION(
+				"UPDATE %PREFIX%SESSION SET LAST_ACCESS_TIME = ?, PRINCIPAL_NAME = ?, SESSION_BYTES = ? " +
+						"WHERE SESSION_ID = ?"),
+
+		DELETE_SESSION(
+				"DELETE FROM %PREFIX%SESSION WHERE SESSION_ID = ?"),
+
+		LIST_SESSIONS_BY_PRINCIPAL_NAME(
+				"SELECT SESSION_BYTES FROM %PREFIX%SESSION WHERE PRINCIPAL_NAME = ?"),
+
+		DELETE_SESSIONS_BY_LAST_ACCESS_TIME(
+				"DELETE FROM %PREFIX%SESSION WHERE LAST_ACCESS_TIME < ?");
+
+		private String sql;
+
+		Query(String sql) {
+			this.sql = sql;
+		}
+
+		public String getSql() {
+			return this.sql;
+		}
+
+	}
+
+	private final JdbcOperations jdbcOperations;
+
+	private final RowMapper<MapSession> mapper = new SessionMapper();
+
+	private final Map<Query, String> queryCache = new HashMap<Query, String>();
+
+	private ApplicationEventPublisher applicationEventPublisher = new ApplicationEventPublisher() {
```

Comment:
If we keep this, need to add `publishEvent` to work with Spring 4.2.x (which we now compile with).